src/amd64/amd64-51-30k-choose_tp.s

```diff
@@ -0,0 +1,2189 @@
+movq   168(%rcx,%rdi),%rbp
+
+# qhasm: tysubx1 = t if =
+# asm 1: cmove <t=int64#15,<tysubx1=int64#6
+# asm 2: cmove <t=%rbp,<tysubx1=%r9
+cmove %rbp,%r9
+
+# qhasm: t = *(uint64 *)(basep + 176 + pos)
+# asm 1: movq   176(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   176(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   176(%rcx,%rdi),%rbp
+
+# qhasm: tysubx2 = t if =
+# asm 1: cmove <t=int64#15,<tysubx2=int64#7
+# asm 2: cmove <t=%rbp,<tysubx2=%rax
+cmove %rbp,%rax
+
+# qhasm: t = *(uint64 *)(basep + 184 + pos)
+# asm 1: movq   184(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   184(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   184(%rcx,%rdi),%rbp
+
+# qhasm: tysubx3 = t if =
+# asm 1: cmove <t=int64#15,<tysubx3=int64#8
+# asm 2: cmove <t=%rbp,<tysubx3=%r10
+cmove %rbp,%r10
+
+# qhasm: t = *(uint64 *)(basep + 192 + pos)
+# asm 1: movq   192(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   192(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   192(%rcx,%rdi),%rbp
+
+# qhasm: tysubx4 = t if =
+# asm 1: cmove <t=int64#15,<tysubx4=int64#9
+# asm 2: cmove <t=%rbp,<tysubx4=%r11
+cmove %rbp,%r11
+
+# qhasm: t = *(uint64 *)(basep + 200 + pos)
+# asm 1: movq   200(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   200(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   200(%rcx,%rdi),%rbp
+
+# qhasm: txaddy0 = t if =
+# asm 1: cmove <t=int64#15,<txaddy0=int64#10
+# asm 2: cmove <t=%rbp,<txaddy0=%r12
+cmove %rbp,%r12
+
+# qhasm: t = *(uint64 *)(basep + 208 + pos)
+# asm 1: movq   208(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   208(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   208(%rcx,%rdi),%rbp
+
+# qhasm: txaddy1 = t if =
+# asm 1: cmove <t=int64#15,<txaddy1=int64#11
+# asm 2: cmove <t=%rbp,<txaddy1=%r13
+cmove %rbp,%r13
+
+# qhasm: t = *(uint64 *)(basep + 216 + pos)
+# asm 1: movq   216(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   216(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   216(%rcx,%rdi),%rbp
+
+# qhasm: txaddy2 = t if =
+# asm 1: cmove <t=int64#15,<txaddy2=int64#12
+# asm 2: cmove <t=%rbp,<txaddy2=%r14
+cmove %rbp,%r14
+
+# qhasm: t = *(uint64 *)(basep + 184 + pos)
+# asm 1: movq   184(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   184(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   224(%rcx,%rdi),%rbp
+
+# qhasm: txaddy3 = t if =
+# asm 1: cmove <t=int64#15,<txaddy3=int64#13
+# asm 2: cmove <t=%rbp,<txaddy3=%r15
+cmove %rbp,%r15
+
+# qhasm: t = *(uint64 *)(basep + 232 + pos)
+# asm 1: movq   232(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   232(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   232(%rcx,%rdi),%rbp
+
+# qhasm: txaddy4 = t if =
+# asm 1: cmove <t=int64#15,<txaddy4=int64#14
+# asm 2: cmove <t=%rbp,<txaddy4=%rbx
+cmove %rbp,%rbx
+
+# qhasm: =? u - 3
+# asm 1: cmp  $3,<u=int64#5
+# asm 2: cmp  $3,<u=%r8
+cmp  $3,%r8
+
+# qhasm: t = *(uint64 *)(basep + 320 + pos)
+# asm 1: movq   320(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   320(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   320(%rcx,%rdi),%rbp
```

Comment:
From this point, the offets change from -40 to -80. Why ?

Reply:
The code is touching every entry in a table (constant time), and conditionally moving the entry related to the bits set in the secret key. Each entry in the table now has an additional `z` field value, along with the 3 original field values, and each field value is 40 bytes (64bits*5limbs / 8bitsbyte). So after every entry in the table, the offset needed for the first original field value increases by 40. 

The [field element struct](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/fe25519.h#L23), the [table entry struct for the original ASM](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/ge25519.h#L61), and the [new table entry struct for this ASM](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/ge25519.h#L68) just below it. The last `struct` has a `z` parameter which allows field inversions to be "batched" once, after the main processing loop. The original never had this because ed25519 signing uses a fixed basepoint (`G`), so the tables were pre-computed with every `z` value normalized to `1` for time/space savings.